agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+type Server struct {
+	logger       *log.Logger
+	cfgMgr       ConfigManager
+	authz        ConnectAuthz
+	resolveToken ACLResolverFunc
+}
+
+// NewServer creates a usable server instance.
+func NewServer(logger *log.Logger, cfgMgr ConfigManager,
+	aclFn ACLResolverFunc, authz ConnectAuthz) *Server {
+	return &Server{
+		logger:       logger,
+		cfgMgr:       cfgMgr,
+		authz:        authz,
+		resolveToken: aclFn,
+	}
+}
+
+// StreamAggregatedResources implements
+// discovery.AggregatedDiscoveryServiceServer. This is the ADS endpoint which is
+// the only xDS API we directly support for now.
+func (s *Server) StreamAggregatedResources(stream ADSStream) error {
+	// a channel for receiving incoming requests
+	reqCh := make(chan *v2.DiscoveryRequest)
+	reqStop := int32(0)
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if atomic.LoadInt32(&reqStop) != 0 {
+				return
+			}
+			if err != nil {
+				close(reqCh)
+				return
+			}
+			reqCh <- req
+		}
+	}()
+
+	err := s.process(stream, reqCh)
+	if err != nil {
+		s.logger.Printf("[DEBUG] Error handling ADS stream: %s", err)
+	}
+
+	// prevents writing to a closed channel if send failed on blocked recv
+	atomic.StoreInt32(&reqStop, 1)
+
+	return err
+}
+
+const (
+	stateInit int = iota
+	statePendingAuth
+	stateRunning
+)
+
+func (s *Server) process(stream ADSStream, reqCh <-chan *v2.DiscoveryRequest) error {
+	// xDS requires a unique nonce to correlate response/request pairs
+	var nonce uint64
+
+	// xDS works with versions of configs. Internally we don't have a consistent
+	// version. We could just hash the config since versions don't have to be
+	// ordered as far as I can tell, but it's cheaper just to increment a counter
+	// every time we observe a new config since the upstream proxycfg package only
+	// delivers updates when there are actual changes.
+	var configVersion uint64
+
+	// Loop state
+	var cfgSnap *proxycfg.ConfigSnapshot
+	var req *v2.DiscoveryRequest
+	var ok bool
+	var stateCh <-chan *proxycfg.ConfigSnapshot
+	var watchCancel func()
+	var proxyID string
+
+	// need to run a small state machine to get through initial authentication.
+	var state = stateInit
+
+	// Configure handlers for each type of request
+	handlers := map[string]*xDSType{
+		EndpointType: &xDSType{
+			typeURL:   EndpointType,
+			resources: endpointsFromSnapshot,
+			stream:    stream,
+		},
+		ClusterType: &xDSType{
+			typeURL:   ClusterType,
+			resources: clustersFromSnapshot,
+			stream:    stream,
+		},
+		RouteType: &xDSType{
+			typeURL:   RouteType,
+			resources: routesFromSnapshot,
+			stream:    stream,
+		},
+		ListenerType: &xDSType{
+			typeURL:   ListenerType,
+			resources: listenersFromSnapshot,
+			stream:    stream,
+		},
+	}
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			// TODO: Not sure if we should error here, this presumably happens during
+			// a normal shutdown where client goes away too.
+			return nil
+		case req, ok = <-reqCh:
+			if !ok {
```

Comment:
Do we even really need to close the channel out and handle this? Or can we just block forever because `stream.Context().Done()` is always set? 

I don't really care, but just wondering because usually its not that critical to close channels unless the receiver needs to really know its closed. But in this case we seem to have an authoritative way to cancel.

Reply:
I'm not sure is the honest answer and was playing it safe.

The check here is because we close this channel if the `Recv` fails on the stream in another goroutine - I'm not sure if we are guaranteed that gRPC will always cancel the stream's context in that case but we should abort processing the RPC as we can't make progress any more.

gRPC guarantees to cancel the stream context once the RPC has returned which we can't observe here since this method blocks the RPC method), I included the case here in case it also does it earlier if the client hangs up or similar. I expect we'd see that via a Recv fail too/instead, just couldn't find clarity on what guarantees are made about the ordering there so just handled everything that might happen!

Even the code in the official XDS server SDK has TODO comments to figure out how gRPC cancels stuff.

So yeah I'd like to spend more time reading gRPC code to get this right, but this is a defensive approach in liu of having time to investigate more 😄.

Reply:
According to https://github.com/grpc/grpc-go/issues/465 it looks like the stream context is only ever cancelled on the server side once the RPC handler returns so the this is actually necessary to be signalled to stop on a Recv error (which is how we know if the client drops).

I'll add a comment and remove the context case here since it seems like it can never be hit now.